bot/slack/slack_bot.py

```diff
@@ -0,0 +1,161 @@
+import random
+import sys
+import os
+import time
+
+from slackclient import SlackClient
+
+from bot_plugin_set import BotPluginSet
+
+from mozdef_util.utilities.logger import logger
+
+
+greetings = [
+    "mozdef bot in da house",
+    "mozdef here..what's up",
+    "mozdef has joined the room..no one panic",
+    "mozdef bot here..nice to see everyone"
+]
+
+
+def formatAlert(alert_dict):
+    summary = alert_dict['summary']
+    if 'category' in alert_dict.keys():
+        summary = "_{0}_: {1}".format(alert_dict['category'], summary)
+    return summary
+
+
+class SlackBot():
+    def __init__(self, api_key, channels, bot_name, async_func):
+        self.slack_client = SlackClient(api_key)
+        self.channels = channels
+        self.bot_name = bot_name
+        self.async_func = async_func
+        self.load_commands()
+
+    def load_commands(self):
+        plugin_dir = os.path.abspath(os.path.join(os.path.dirname(__file__), 'commands'))
+        self.plugin_set = BotPluginSet(plugin_dir)
+
+    def run(self):
+        if self.slack_client.rtm_connect():
+            logger.info("Bot connected to slack")
+            self.post_welcome_message(random.choice(greetings))
+            self.async_func(self)
+            self.listen_for_messages()
+        else:
+            logger.error("Unable to connect to slack")
+            sys.exit(1)
+
+    def handle_command(self, message_text):
+        response = ""
+        message_tokens = message_text.split()
+        command = message_tokens[0]
+        parameters = message_tokens[1:len(message_tokens)]
+
+        if command == '!help':
+            response = "\nHelp is on it's way...try these:\n"
+            for plugin in self.plugin_set.enabled_plugins:
+                response += "\n{0} -- {1}".format(
+                    plugin['command_name'],
+                    plugin['help_text']
+                )
+        else:
+            # Loop through plugins and send message to it
+            plugin_command_found = False
+            for plugin in self.plugin_set.enabled_plugins:
+                if command == plugin['command_name']:
+                    plugin_command_found = True
+                    response += "\n" + plugin['plugin_class'].handle_command(parameters)
+                    continue
+
+            if plugin_command_found is False:
+                response = "Unknown command: " + command + ". Try !help"
+        return response
+
+    def parse_command(self, content):
```

Comment:
I'm not sure that this approach will serve us well in the long term.

It wouldn't take much more effort to build parsing/tokenizing logic into each command. This way, the SlackBot only has to be able to identify commands by the first "word" in a command string, and can delegate the work of making sense of tokens to the command itself.

Reply:
This is actually how it's currently setup, we tokenize the entire slack message into a command + parameters (https://github.com/mozilla/MozDef/pull/1018/files#diff-46fed7650379c8a8432b6efca3b11a91R53 and https://github.com/mozilla/MozDef/pull/1018/files#diff-46fed7650379c8a8432b6efca3b11a91R54). We then identify which plugin (if any) is associated with the command, and simply pass the plugin the parameters (https://github.com/mozilla/MozDef/pull/1018/files#diff-46fed7650379c8a8432b6efca3b11a91R69).

Is this the functionality you were requesting, or am I misunderstanding the comment?

Reply:
Oh, upon second thought, do you mean to have each plugin be the one to tokenize the parameters? So if the slack message is "!ipinfo 1.2.3.4 5.6.7.8", the parameter we would pass to the plugin would be "1.2.3.4 5.6.7.8" (instead of ['1.2.3.4', '5.6.7.8'] with how it's currently implemented)?

Reply:
I see what you mean. My suggestion might not be necessary.